From 4a66db1531c18f64b86a44bb6cb8da904c50ac7b Mon Sep 17 00:00:00 2001
From: Tom Marshall <tdm@cyngn.com>
Date: Thu, 5 Nov 2015 11:20:54 -0800
Subject: [PATCH 05/18] vold: Honor mount options for ext4 partitions

 * Save mount options and fstype for DiskPartition objects
 * Pass mount options and fstype to PublicVolume ctor
 * Pass mount options to Ext4::Mount
 * Use specified/default fstype if blkid fails

This is trivially extensible for other fs types.

Change-Id: Ie5dd27f26a4a7129dd9efe6a7ad0a589928282a0
---
 Disk.cpp          |  6 ++++--
 Disk.h            |  4 +++-
 DiskPartition.cpp |  9 ++++++---
 DiskPartition.h   |  7 ++++++-
 PublicVolume.cpp  |  9 ++++++---
 PublicVolume.h    |  5 ++++-
 VolumeManager.cpp |  3 ++-
 VolumeManager.h   | 12 ++++++++++--
 fs/Ext4.cpp       |  5 +++--
 fs/Ext4.h         |  2 +-
 main.cpp          | 11 ++++++++++-
 11 files changed, 55 insertions(+), 18 deletions(-)

diff --git a/Disk.cpp b/Disk.cpp
index aa8750f..5316711 100644
--- a/Disk.cpp
+++ b/Disk.cpp
@@ -164,8 +164,10 @@ status_t Disk::destroy() {
     return OK;
 }
 
-void Disk::createPublicVolume(dev_t device) {
-    auto vol = std::shared_ptr<VolumeBase>(new PublicVolume(device, mNickname));
+void Disk::createPublicVolume(dev_t device,
+                const std::string& fstype /* = "" */,
+                const std::string& mntopts /* = "" */) {
+    auto vol = std::shared_ptr<VolumeBase>(new PublicVolume(device, mNickname, fstype, mntopts));
     if (mJustPartitioned) {
         LOG(DEBUG) << "Device just partitioned; silently formatting";
         vol->setSilent(true);
diff --git a/Disk.h b/Disk.h
index c94e829..a334069 100644
--- a/Disk.h
+++ b/Disk.h
@@ -110,7 +110,9 @@ protected:
     /* Flag that we just partitioned and should format all volumes */
     bool mJustPartitioned;
 
-    void createPublicVolume(dev_t device);
+    void createPublicVolume(dev_t device,
+                    const std::string& fstype = "",
+                    const std::string& mntopts = "");
     void createPrivateVolume(dev_t device, const std::string& partGuid);
 
     void destroyAllVolumes();
diff --git a/DiskPartition.cpp b/DiskPartition.cpp
index 262d0b7..662118c 100644
--- a/DiskPartition.cpp
+++ b/DiskPartition.cpp
@@ -44,9 +44,12 @@ namespace android {
 namespace vold {
 
 DiskPartition::DiskPartition(const std::string& eventPath, dev_t device,
-            const std::string& nickname, int flags, int partnum) :
+            const std::string& nickname, int flags, int partnum,
+            const std::string& fstype /* = "" */, const std::string& mntopts /* = "" */) :
         Disk(eventPath, device, nickname, flags),
-        mPartNum(partnum) {
+        mPartNum(partnum),
+        mFsType(fstype),
+        mMntOpts(mntopts) {
     // Empty
 }
 
@@ -59,7 +62,7 @@ status_t DiskPartition::create() {
     mCreated = true;
     notifyEvent(ResponseCode::DiskCreated, StringPrintf("%d", mFlags));
     dev_t partDevice = makedev(major(mDevice), minor(mDevice) + mPartNum);
-    createPublicVolume(partDevice);
+    createPublicVolume(partDevice, mFsType, mMntOpts);
     return OK;
 }
 
diff --git a/DiskPartition.h b/DiskPartition.h
index cf440a6..b1fa33b 100644
--- a/DiskPartition.h
+++ b/DiskPartition.h
@@ -31,7 +31,8 @@ class DiskPartition : public Disk {
 public:
     DiskPartition(const std::string& eventPath, dev_t device,
             const std::string& nickname,
-            int flags, int partnum);
+            int flags, int partnum,
+            const std::string& fstype = "", const std::string& mntopts = "");
     virtual ~DiskPartition();
 
     virtual status_t create();
@@ -44,6 +45,10 @@ public:
 private:
     /* Partition number */
     int mPartNum;
+    /* Filesystem type */
+    std::string mFsType;
+    /* Mount options */
+    std::string mMntOpts;
 };
 
 }  // namespace vold
diff --git a/PublicVolume.cpp b/PublicVolume.cpp
index cb9e102..0b47548 100644
--- a/PublicVolume.cpp
+++ b/PublicVolume.cpp
@@ -49,8 +49,11 @@ static const char* kFusePath = "/system/bin/sdcard";
 
 static const char* kAsecPath = "/mnt/secure/asec";
 
-PublicVolume::PublicVolume(dev_t device, const std::string& nickname) :
-        VolumeBase(Type::kPublic), mDevice(device), mFusePid(0), mFsLabel(nickname) {
+PublicVolume::PublicVolume(dev_t device, const std::string& nickname,
+                const std::string& fstype /* = "" */,
+                const std::string& mntopts /* = "" */) :
+        VolumeBase(Type::kPublic), mDevice(device), mFusePid(0),
+        mFsType(fstype), mFsLabel(nickname), mMntOpts(mntopts) {
     setId(StringPrintf("public:%u,%u", major(device), minor(device)));
     mDevPath = StringPrintf("/dev/block/vold/%s", getId().c_str());
 }
@@ -164,7 +167,7 @@ status_t PublicVolume::doMount() {
         ret = exfat::Mount(mDevPath, mRawPath, false, false, false,
                 AID_MEDIA_RW, AID_MEDIA_RW, 0007);
     } else if (mFsType == "ext4") {
-        ret = ext4::Mount(mDevPath, mRawPath, false, false, true);
+        ret = ext4::Mount(mDevPath, mRawPath, false, false, true, mMntOpts);
     } else if (mFsType == "f2fs") {
         ret = f2fs::Mount(mDevPath, mRawPath);
     } else if (mFsType == "ntfs") {
diff --git a/PublicVolume.h b/PublicVolume.h
index 08da34e..e9e2926 100644
--- a/PublicVolume.h
+++ b/PublicVolume.h
@@ -39,7 +39,8 @@ namespace vold {
  */
 class PublicVolume : public VolumeBase {
 public:
-    explicit PublicVolume(dev_t device, const std::string& nickname);
+    PublicVolume(dev_t device, const std::string& nickname,
+                    const std::string& mntopts = "", const std::string& fstype = "");
     virtual ~PublicVolume();
 
 protected:
@@ -73,6 +74,8 @@ private:
     std::string mFsUuid;
     /* User-visible filesystem label */
     std::string mFsLabel;
+    /* Mount options */
+    std::string mMntOpts;
 
     DISALLOW_COPY_AND_ASSIGN(PublicVolume);
 };
diff --git a/VolumeManager.cpp b/VolumeManager.cpp
index d0f4605..6c4771c 100644
--- a/VolumeManager.cpp
+++ b/VolumeManager.cpp
@@ -316,7 +316,8 @@ void VolumeManager::handleBlockEvent(NetlinkEvent *evt) {
                                 source->getNickname(), flags) :
                         new android::vold::DiskPartition(eventPath, device,
                                 source->getNickname(), flags,
-                                source->getPartNum());
+                                source->getPartNum(),
+                                source->getFsType(), source->getMntOpts());
                 disk->create();
                 mDisks.push_back(std::shared_ptr<android::vold::Disk>(disk));
                 break;
diff --git a/VolumeManager.h b/VolumeManager.h
index c43aae6..dfd8c8f 100644
--- a/VolumeManager.h
+++ b/VolumeManager.h
@@ -97,8 +97,12 @@ public:
 
     class DiskSource {
     public:
-        DiskSource(const std::string& sysPattern, const std::string& nickname, int partnum, int flags) :
-                mSysPattern(sysPattern), mNickname(nickname), mPartNum(partnum), mFlags(flags) {
+        DiskSource(const std::string& sysPattern, const std::string& nickname,
+                        int partnum, int flags,
+                        const std::string& fstype, const std::string mntopts) :
+                mSysPattern(sysPattern), mNickname(nickname),
+                mPartNum(partnum), mFlags(flags),
+                mFsType(fstype), mMntOpts(mntopts) {
         }
 
         bool matches(const std::string& sysPath) {
@@ -108,12 +112,16 @@ public:
         const std::string& getNickname() { return mNickname; }
         int getPartNum() { return mPartNum; }
         int getFlags() { return mFlags; }
+        const std::string& getFsType() { return mFsType; }
+        const std::string& getMntOpts() { return mMntOpts; }
 
     private:
         std::string mSysPattern;
         std::string mNickname;
         int mPartNum;
         int mFlags;
+        std::string mFsType;
+        std::string mMntOpts;
     };
 
     void addDiskSource(const std::shared_ptr<DiskSource>& diskSource);
diff --git a/fs/Ext4.cpp b/fs/Ext4.cpp
index 5c202f3..b254279 100644
--- a/fs/Ext4.cpp
+++ b/fs/Ext4.cpp
@@ -133,12 +133,13 @@ status_t Check(const std::string& source, const std::string& target) {
 }
 
 status_t Mount(const std::string& source, const std::string& target, bool ro,
-        bool remount, bool executable) {
+        bool remount, bool executable, const std::string& opts /* = "" */) {
     int rc;
     unsigned long flags;
 
     const char* c_source = source.c_str();
     const char* c_target = target.c_str();
+    const char* c_opts = opts.c_str();
 
     flags = MS_NOATIME | MS_NODEV | MS_NOSUID | MS_DIRSYNC;
 
@@ -146,7 +147,7 @@ status_t Mount(const std::string& source, const std::string& target, bool ro,
     flags |= (ro ? MS_RDONLY : 0);
     flags |= (remount ? MS_REMOUNT : 0);
 
-    rc = mount(c_source, c_target, "ext4", flags, NULL);
+    rc = mount(c_source, c_target, "ext4", flags, c_opts);
 
     if (rc && errno == EROFS) {
         SLOGE("%s appears to be a read only filesystem - retrying mount RO", c_source);
diff --git a/fs/Ext4.h b/fs/Ext4.h
index f78dc95..75db4f1 100644
--- a/fs/Ext4.h
+++ b/fs/Ext4.h
@@ -29,7 +29,7 @@ bool IsSupported();
 
 status_t Check(const std::string& source, const std::string& target);
 status_t Mount(const std::string& source, const std::string& target, bool ro,
-        bool remount, bool executable);
+        bool remount, bool executable, const std::string& opts = "");
 status_t Format(const std::string& source, unsigned long numSectors,
         const std::string& target);
 status_t Resize(const std::string& source, unsigned long numSectors);
diff --git a/main.cpp b/main.cpp
index e5e6d1e..18d5375 100644
--- a/main.cpp
+++ b/main.cpp
@@ -222,6 +222,14 @@ static int process_config(VolumeManager *vm) {
     for (int i = 0; i < fstab->num_entries; i++) {
         if (fs_mgr_is_voldmanaged(&fstab->recs[i])) {
             std::string sysPattern(fstab->recs[i].blk_device);
+            std::string fstype;
+            if (fstab->recs[i].fs_type) {
+                fstype = fstab->recs[i].fs_type;
+            }
+            std::string mntopts;
+            if (fstab->recs[i].fs_options) {
+                mntopts = fstab->recs[i].fs_options;
+            }
             std::string nickname(fstab->recs[i].label);
             int partnum = fstab->recs[i].partnum;
             int flags = 0;
@@ -239,7 +247,8 @@ static int process_config(VolumeManager *vm) {
             }
 
             vm->addDiskSource(std::shared_ptr<VolumeManager::DiskSource>(
-                    new VolumeManager::DiskSource(sysPattern, nickname, partnum, flags)));
+                    new VolumeManager::DiskSource(sysPattern, nickname, partnum, flags,
+                                    fstype, mntopts)));
         }
     }
     property_set("vold.has_adoptable", has_adoptable ? "1" : "0");
-- 
2.8.2

